### PR TITLE
[CALCITE-5510] Support sort by ordinal when sort by column is an expression in RelToSqlConverter

### PR DESCRIPTION
**Pull request by peng128**
Keep the output sort by column to ordinary when sort by ordinal in dialect is true and the column is an expression.

### CODE REVIEW

**Comment by zoudan** [comment]
LGTM

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java
Comment (chunweilei) [inline core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:765]: COUNT(*) is replaced by 2 rather than 1. So the ordinal starts from 1?

Comment (peng128) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:765]: Yes, I think the ordinal is the position of column in select list, it starts from 1.

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java
Comment (libenchao) [inline core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:820]: Is `order by 3 is null` valid here? In [Calcite](https://github.com/apache/calcite/blob/56e9deff51942fe7673096aaf5f50544f47f1411/core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java#L6677-L6680), we'll take this not a 'field reference'. I'm not sure how other databases behaves in this case.

File: core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java
Comment (libenchao) [inline core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:620]: `filed` -> `field`

**Review by libenchao — COMMENTED** [review]
@peng128 thanks for your PR, I've left some comments.

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java
Comment (peng128) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:820]: I think it valid here.
I try to make a test to compare the relNode compare '3 is null' sql with the original SQL and the 'count(*) is null' sql.
Using default SqlParser config.

1. For original sql:
```
select "product_class_id", "brand_name",count(*) as c
from "product"
group by rollup("product_class_id", "brand_name")
order by 1, 2, 3
```
relNode is
```
LogicalSort(sort0=[$0], sort1=[$1], sort2=[$2], dir0=[ASC], dir1=[ASC], dir2=[ASC])
  LogicalAggregate(group=[{0, 1}], groups=[[{0, 1}, {0}, {}]], C=[COUNT()])
    LogicalProject(product_class_id=[$0], brand_name=[$2])
      JdbcTableScan(table=[[foodmart, product]])
```


2. For 'count(*) is null' sql:
```
select "product_class_id", "brand_name", count(*) as c
from "product"
group by rollup("product_class_id", "brand_name")
order by "product_class_id", "brand_name", count(*) is null, count(*)
```
relNode is
```
LogicalProject(product_class_id=[$0], brand_name=[$1], C=[$2])
  LogicalSort(sort0=[$0], sort1=[$1], sort2=[$3], sort3=[$2], dir0=[ASC], dir1=[ASC], dir2=[ASC], dir3=[ASC])
    LogicalProject(product_class_id=[$0], brand_name=[$1], C=[$2], EXPR$3=[false])
      LogicalAggregate(group=[{0, 1}], groups=[[{0, 1}, {0}, {}]], C=[COUNT()])
        LogicalProject(product_class_id=[$0], brand_name=[$2])
          JdbcTableScan(table=[[foodmart, product]])
```


3. For '3 is null' sql:
```
select "product_class_id", "brand_name", count(*) as c
from "product"
group by rollup("product_class_id", "brand_name")
order by 1, 2, 3 is null, 3
```
relNode is
```
LogicalProject(product_class_id=[$0], brand_name=[$1], C=[$2])
  LogicalSort(sort0=[$0], sort1=[$1], sort2=[$3], sort3=[$2], dir0=[ASC], dir1=[ASC], dir2=[ASC], dir3=[ASC])
    LogicalProject(product_class_id=[$0], brand_name=[$1], C=[$2], EXPR$3=[false])
      LogicalAggregate(group=[{0, 1}], groups=[[{0, 1}, {0}, {}]], C=[COUNT()])
        LogicalProject(product_class_id=[$0], brand_name=[$2])
          JdbcTableScan(table=[[foodmart, product]])
```

The relNode of '3 is null' sql and 'count(*) is null' sql is almost some. Is my understanding correct?
So I think it's fine here. And I test '3 is null' sql in MySQL, it works.

File: core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java
Comment (peng128) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java]: Fixed it. Thanks~

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java
Comment (libenchao) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:820]: No, what I mean it that we only takes pure literal as a field reference in Calcite, if it's an expression, it won't be taken as a field reference. Your second case shows what I described above.

To be more precise, `3 is null` equals `false`, not `count(*) is null` in this case.

I'm not saying Calcite is definitely correct for current behavior. We need to investigate what other databases behave when the literal is in a expression, and let Calcite be consistent with the majority of them.

If we prefer to not take literal in expressions as a field reference, then your change now would make it incorrect.

Comment (peng128) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:820]: I got it. We shouldn't take literal in experssion as field reference.
How about keep original expression when we need to sort null values in order by?
For example, the output sql will be 
```
select "product_class_id", "brand_name", count(*) as c
from "product"
group by rollup("product_class_id", "brand_name")
order by 1, 2, count(*) is null, 3
```

Comment (libenchao) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:820]: This sounds good to me.

Comment (peng128) [reply core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java]: I fixed it. Please review again, thanks.

File: core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java
Comment (libenchao) [inline core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:1765]: `final`

**Review by libenchao — COMMENTED** [review]
Looks good to me now, with one minor comment. After fixing that, I'll merge this PR.

**Comment by libenchao** [comment]
reopened, because slow tests fail due to this change: https://github.com/apache/calcite/actions/runs/4223301074/jobs/7332876394

File: core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java
Comment (libenchao) [inline core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:1824]: The [last version we merged](https://github.com/apache/calcite/commit/0e8d6b093d26c291361bfd63ecb14828d1e33bc6) does not have this logic, could you help me to understand why do you need to add this logic now?

Comment (peng128) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:1824]: We found test case LatticeTest#testSum is failed in slow test.
These code is to fix it.

The reason of test case failure is that the project will be merged when the sub-relnode is sort(change by [CALCITE-4901](https://issues.apache.org/jira/browse/CALCITE-4901)).

For example,
If the relNode is
```
JdbcProject(C=[$1])
  JdbcSort(sort0=[$1], dir0=[DESC], fetch=[1])
    JdbcAggregate(group=[{0}], C=[$SUM0($7)])
      JdbcTableScan(table=[[foodmart, sales_fact_1997]])
```
The result after merge project is
```
SELECT SUM(`sales_fact_1997`.`unit_sales`) AS `C`
FROM `foodmart`.`sales_fact_1997` AS `sales_fact_1997`
GROUP BY `sales_fact_1997`.`product_id`
ORDER BY 2 DESC
```
it's a incorrect sql, we want
```
SELECT `C` FROM (
SELECT `sales_fact_1997`.`product_id`, SUM(`sales_fact_1997`.`unit_sales`) AS `C`
FROM `foodmart`.`sales_fact_1997` AS `sales_fact_1997`
GROUP BY `sales_fact_1997`.`product_id`
ORDER BY 2 DESC) t
```
So, these code to find out whether there is sort by original under project. 
And don't merge project when there is sort by original under project.

Comment (libenchao) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:1824]: Since we agreed that the slow tests failure is not caused by this PR, we'd better handle this in CALCITE-5537.

Comment (peng128) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:1824]: There are three test cases failure.
1. LatticeTest#testSum
2. LatticeSuggesterTest#testFoodMartAll
3. LatticeSuggesterTest#testFoodMartAllEvolve

Case 1 is caused by this PR. Maybe we could fix in this PR?
Case 2 & 3 is not caused by this PR. I think we could fix case 2 & 3 in [CALCITE-5537](https://issues.apache.org/jira/browse/CALCITE-5537).

BTW, these code just fix case 1.

Comment (libenchao) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:1824]: Thanks for the clarification, I agree with you. I'll review the PR in the coming days.

**Review by libenchao — APPROVED** [review]
+1, merging. Thanks @peng128 for your contribution!

**Pull request closed without merging**
